js-packages/@prestojs/viewmodel/src/useViewModelChangeEffect.ts

```diff
@@ -0,0 +1,174 @@
+// @flow
+import { useEffect, useRef } from 'react';
+import ViewModel from './ViewModel';
+
+export const ADD = 'ADD';
+export const UPDATE = 'UPDATE';
+export const DELETE = 'DELETE';
+
+type ChangeType = 'ADD' | 'UPDATE' | 'DELETE';
+
+type Options = {
+    /**
+     * If true then when any records are added it will call `run` with 'ADD'
+     * as the first parameter.
+     */
+    runOnAdd?: boolean;
+    /**
+     * If true then when any records are deleted it will call `run` with 'DELETE'
+     * as the first parameter.
+     */
+    runOnDelete?: boolean;
+    /**
+     * If true then when any records are updated it will call `run` with 'UPDATE'
+     * as the first parameter.
+     */
+    runOnUpdate?: boolean;
+    /** If true will pass through diff object to `run` on second parameter */
+    includeDiff?: boolean;
+};
+
+type AddedDiff<T extends ViewModel> = {
+    previous: T[];
+    added: T[];
+};
+
+type UpdatedDiff<T extends ViewModel> = {
+    previous: T[];
+    updated: T[];
+};
+
+type DeletedDiff<T extends ViewModel> = {
+    previous: T[];
+    deleted: T[];
+};
+
+type Diff<T extends ViewModel> = AddedDiff<T> | UpdatedDiff<T> | DeletedDiff<T>;
+
+function isEqual<T extends ViewModel>(a: T[], b: T[]): boolean {
+    if (a.length !== b.length) {
+        return false;
+    }
+    for (const i in a) {
+        if (!a[i].isEqual(b[i])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Run an effect after a collection of records changes.
+ *
+ * This works by monitoring the collection of records passed in - it does
+ * _not_ listen to changes in a cache. As such make sure you only pass in
+ * latest version of records from a cache (eg. by using useViewModelCache).
+ * If you pass in records that are cached on your component then no changes
+ * in the cache will be detected.
+ *
+ * @param records The records to monitor for changes. This should usually
+ * come from a `useViewModelCache` or another hook that returns data live
+ * from he cache. If a falsy value is passed then no changes will be detected.
+ * Use this to opt out of checks (for example when a fetch is in progress).
+ * @param run The function to call when `records` changes. The first parameter
+ * will be one of 'ADD', 'UPDATE' or 'DELETE' depending on the change. If
+ * `options.includeDiff` is true then the second parameter will be an object
+ * describing the change with the shape:
+ *
```

Comment:
`includeDiff` is not set by default since it involves extra processing? Does this need to be mentioned so that devs don't do it out of habit or is it actually not that expensive?

Reply:
yeah it can be expensive, i can add a note

Reply:
I just double-checked and couldn't find this (I might have missed it)